qiskit/aqua/operators/gradients/circuit_qfis/overlap_diag.py

```
@@ -0,0 +1,258 @@
+        """
+        Args:
+            operator: The operator corresponding to the quantum state |ψ(ω)〉for which we compute
+                the QFI
+            params: The parameters we are computing the QFI wrt: ω
+
+        Returns:
+            ListOp where the operator at position k corresponds to QFI_k,k
+
+        Raises:
+            NotImplementedError: If a circuit is found such that one parameter controls multiple
+                gates, or one gate contains multiple parameters.
+
+        """
+
+        if not isinstance(operator, (CircuitOp, CircuitStateFn)):
+            raise NotImplementedError('operator must be a CircuitOp or CircuitStateFn')
+
+        circuit = operator.primitive
+
+        # Partition the circuit into layers, and build the circuits to prepare $\psi_i$
+        layers = self._partition_circuit(circuit)
+        if layers[-1].num_parameters == 0:
+            layers.pop(-1)
+
+        psis = [CircuitOp(layer) for layer in layers]
+        for i, psi in enumerate(psis):
+            if i == 0:
+                continue
+            psis[i] = psi @ psis[i - 1]
+
+        # TODO: make this work for other types of rotations
+        # NOTE: This assumes that each parameter only affects one rotation.
+        # we need to think more about what happens if multiple rotations
+        # are controlled with a single parameter.
+        generators = self._get_generators(params, circuit)
+
+        diag = []
+        for param in params:
+            if len(circuit._parameter_table[param]) > 1:
+                raise NotImplementedError("OverlapDiag does not yet support multiple "
+                                          "gates parameterized by a single parameter. For such "
+                                          "circuits use LinCombFull")
+
+            gate = circuit._parameter_table[param][0][0]
+
+            assert len(gate.params) == 1, "OverlapDiag cannot yet support gates with more than " \
```

This should raise a proper error, not just be an assert statement 🙂